internals/controllers/namespace/namespace_sync.go

```diff
@@ -0,0 +1,88 @@
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// sync is being called every time there is an update in the namespace and makes sure its role is up-to-date
+func (r *NamespaceReconciler) sync(nsObject *utils.ObjectContext) error {
+	ctx := nsObject.Ctx
+	logger := log.FromContext(ctx)
+	logger.Info("syncing namespace")
+
+	nsName := nsObject.Object.GetName()
+
+	if err := createNamespaceHNSView(nsObject); err != nil {
+		return fmt.Errorf("failed to create role and roleBinding objects associated with namespace '%s': "+err.Error(), nsName)
+	}
+	logger.Info("successfully created role and roleBinding objects associated with namespace", "namespace", nsName)
+
+	if utils.IsChildlessNamespace(nsObject) {
+		if err := updateNSRole(nsObject, danav1.Leaf); err != nil {
+			return fmt.Errorf("failed to update role of namespace '%s': "+err.Error(), nsName)
+		}
+	} else if err := updateNSRole(nsObject, danav1.NoRole); err != nil {
+		return fmt.Errorf("failed to update role of namespace '%s': "+err.Error(), nsName)
+	}
+	logger.Info("successfully updated role of namespace", "namespace", nsName)
+
+	if err := ensureChildrenSNSResourcePoolLabel(nsObject); err != nil {
+		return fmt.Errorf("failed to set ResourcePool labels of children subnamespaces of namespace '%s': "+err.Error(), nsName)
+	}
+	logger.Info("successfully set ResourcePool labels of children subnamespaces of namespace", "namespace", nsName)
+
+	return nil
+}
+
+// updateNSRole updates the role of a namespace
+func updateNSRole(namespace *utils.ObjectContext, role string) error {
+	return namespace.UpdateObject(func(object client.Object, log logr.Logger) (client.Object, logr.Logger) {
+		log = log.WithValues("updated role annotation", role)
+		object.(*corev1.Namespace).Labels[danav1.Role] = role
+		object.(*corev1.Namespace).Annotations[danav1.Role] = role
+		return object, log
+	})
+}
+
+// ensureChildrenSNSResourcePoolLabel makes sure that the ResourcePool label on the children
+// subnamespaces of a namespace are set correctly, according to the namespace
+func ensureChildrenSNSResourcePoolLabel(nsObject *utils.ObjectContext) error {
+	nsName := nsObject.Object.GetName()
+
+	snsList, err := utils.NewObjectContextList(nsObject.Ctx, nsObject.Client, &danav1.SubnamespaceList{}, client.InNamespace(nsName))
+	if err != nil {
+		return err
+	}
+
+	isNSRP := utils.GetNamespaceResourcePool(nsObject)
+	for _, sns := range snsList.Objects.(*danav1.SubnamespaceList).Items {
+		snsName := sns.GetName()
+		snsObj, err := utils.NewObjectContext(nsObject.Ctx, nsObject.Client, types.NamespacedName{Namespace: nsName, Name: snsName}, &danav1.Subnamespace{})
+		if err != nil {
+			return err
+		}
+
+		// if the child subnamespace is not an upper-rp AND is not the same type as its parent
+		// (for instance one is a ResourcePool and the other is not), then set the type of the child
+		// subnamespace to be the same as its parent
+		isSNSUpperRp, err := utils.IsUpperResourcePool(snsObj)
```

Comment:
move it different function pass all the relevant vars ensure a single children and use it in a loop

Reply:
changed